polkadot/node/network/gossip-support/src/lib.rs

```diff
@@ -271,8 +271,30 @@ where
 					)
 					.await?;
 				}
-				// authority_discovery is just a cache so let's try every leaf to detect if there
-				// are new authorities there.
+				// authority_discovery is just a cache so let's try every time we try to re-connect
+				// if new authorities are present.
+				self.update_authority_ids(sender, session_info.discovery_keys).await;
+			} else if let Some(session_index) = self
+				.last_session_index
+				.filter(|last_session_index| *last_session_index >= current_index)
+			{
+				// authority_discovery is just a cache so let's try every leaf from the current
+				// session if there are new authorities detected and inform the needed subsystems.
+				let session_info =
+					util::request_session_info(leaf, session_index, sender).await.await??;
```

Comment:
why do we need an if let here? shouldn't we just use `else {` and `current_index` since we're using the leaf?

Reply:
Yeah, that would've worked definitely, but I've tried to be defensive here, for the cases where maybe we process late a leaf from the previous session and I did not want to go backwards here.

Reply:
hmm, yeah, i was thinking in that scenario this if let would succeed, but the session info request would not. But actually, it might, because we cache session per session_index regardless of relay_parent.

Reply:
Why wouldn't it succeed ?

Reply:
because we would try to request new session with old leaf (but that will succeed because it will served from cache)